In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder

In [2]:
df = pd.read_csv("train_u6lujuX_CVtuZ9i (1).csv")

In [3]:
df.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 614 entries, 0 to 613
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Loan_ID            614 non-null    object 
 1   Gender             601 non-null    object 
 2   Married            611 non-null    object 
 3   Dependents         599 non-null    object 
 4   Education          614 non-null    object 
 5   Self_Employed      582 non-null    object 
 6   ApplicantIncome    614 non-null    int64  
 7   CoapplicantIncome  614 non-null    float64
 8   LoanAmount         592 non-null    float64
 9   Loan_Amount_Term   600 non-null    float64
 10  Credit_History     564 non-null    float64
 11  Property_Area      614 non-null    object 
 12  Loan_Status        614 non-null    object 
dtypes: float64(4), int64(1), object(8)
memory usage: 62.5+ KB


In [5]:
df.describe()

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
count       614.000000         614.000000  592.000000         600.00000   
mean       5403.459283        1621.245798  146.412162         342.00000   
std        6109.041673        2926.248369   85.587325          65.12041   
min         150.000000           0.000000    9.000000          12.00000   
25%        2877.500000           0.000000  100.000000         360.00000   
50%        3812.500000        1188.500000  128.000000         360.00000   
75%        5795.000000        2297.250000  168.000000         360.00000   
max       81000.000000       41667.000000  700.000000         480.00000   

       Credit_History  
count      564.000000  
mean         0.842199  
std          0.364878  
min          0.000000  
25%          1.000000  
50%          1.000000  
75%          1.000000  
max          1.000000

In [6]:
df.shape

(614, 13)

<p style="font-size:20px; font-weight:bold;">1. Data Preprocessing</p>

<p style="font-size:18px; font-weight:bold;">- Handle missing values</p>

In [7]:
df.duplicated().sum()  

np.int64(0)

In [8]:
df['Gender'] = df['Gender'].fillna(df['Gender'].mode()[0])

In [9]:
df['Married'] = df['Married'].fillna(df['Married'].mode()[0])

In [10]:
df['Dependents'] = df['Dependents'].fillna(df['Dependents'].mode()[0])

In [11]:
df['Self_Employed'] = df['Self_Employed'].fillna(df['Self_Employed'].mode()[0])

In [12]:
df['Credit_History'] = df['Credit_History'].fillna(df['Credit_History'].mode()[0])

In [13]:
df['LoanAmount'] = df['LoanAmount'].fillna(df['LoanAmount'].median())

In [14]:
df['Loan_Amount_Term'] = df['Loan_Amount_Term'].fillna(df['Loan_Amount_Term'].mode()[0])

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 614 entries, 0 to 613
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Loan_ID            614 non-null    object 
 1   Gender             614 non-null    object 
 2   Married            614 non-null    object 
 3   Dependents         614 non-null    object 
 4   Education          614 non-null    object 
 5   Self_Employed      614 non-null    object 
 6   ApplicantIncome    614 non-null    int64  
 7   CoapplicantIncome  614 non-null    float64
 8   LoanAmount         614 non-null    float64
 9   Loan_Amount_Term   614 non-null    float64
 10  Credit_History     614 non-null    float64
 11  Property_Area      614 non-null    object 
 12  Loan_Status        614 non-null    object 
dtypes: float64(4), int64(1), object(8)
memory usage: 62.5+ KB


<p style="font-size:18px; font-weight:bold;">- Encode categorical features</p>

In [16]:
le = LabelEncoder()

In [17]:
df['Gender'] = le.fit_transform(df['Gender'])
df['Married'] = le.fit_transform(df['Married'])
df['Education'] = le.fit_transform(df['Education'])
df['Self_Employed'] = le.fit_transform(df['Self_Employed'])
df['Loan_Status'] = le.fit_transform(df['Loan_Status'])
df['Property_Area'] = le.fit_transform(df['Property_Area'])

In [18]:
df.head()

Loan_ID  Gender  Married Dependents  Education  Self_Employed  \
0  LP001002       1        0          0          0              0   
1  LP001003       1        1          1          0              0   
2  LP001005       1        1          0          0              1   
3  LP001006       1        1          0          1              0   
4  LP001008       1        0          0          0              0   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0       128.0             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History  Property_Area  Loan_Status  
0             1.0              2            1  
1             1.0              0            0  
2             1.0              2            1  
3             1.0              2            1  
4             1.0              2            1

In [19]:
df.dtypes

Loan_ID               object
Gender                 int64
Married                int64
Dependents            object
Education              int64
Self_Employed          int64
ApplicantIncome        int64
CoapplicantIncome    float64
LoanAmount           float64
Loan_Amount_Term     float64
Credit_History       float64
Property_Area          int64
Loan_Status            int64
dtype: object

In [20]:
df['Dependents'].unique()

array(['0', '1', '2', '3+'], dtype=object)

In [21]:
df['Dependents'] = df['Dependents'].replace('3+', 3)
df['Dependents'] = df['Dependents'].astype(float) 

In [22]:
df.dtypes

Loan_ID               object
Gender                 int64
Married                int64
Dependents           float64
Education              int64
Self_Employed          int64
ApplicantIncome        int64
CoapplicantIncome    float64
LoanAmount           float64
Loan_Amount_Term     float64
Credit_History       float64
Property_Area          int64
Loan_Status            int64
dtype: object

<p style="font-size:18px; font-weight:bold;">- Apply feature scaling</p>

In [23]:
from sklearn.preprocessing import StandardScaler

# Scaling numerical columns only
num_cols = ['ApplicantIncome', 'CoapplicantIncome', 'LoanAmount', 'Loan_Amount_Term']

# Initialize the scaler
scaler = StandardScaler()

# Fit and transform the numerical features
df[num_cols] = scaler.fit_transform(df[num_cols])

# Check the scaled values
df[num_cols].head()

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term
0         0.072991          -0.554487   -0.211241          0.273231
1        -0.134412          -0.038732   -0.211241          0.273231
2        -0.393747          -0.554487   -0.948996          0.273231
3        -0.462062           0.251980   -0.306435          0.273231
4         0.097728          -0.554487   -0.056551          0.273231

<p style="font-size:18px; font-weight:bold;">- Split the dataset into training and test sets (70/30)</p>

In [24]:
# Separate features and target
X = df.drop(['Loan_Status','Loan_ID'], axis=1)
y = df['Loan_Status']

In [25]:
# Split into training and test sets (70 / 30)
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, 
    test_size=0.3, 
    random_state=42,
    stratify=y
)

In [26]:
# Check the shape of the split
print("Training set shape:", X_train.shape)
print("Test set shape:", X_test.shape)

Training set shape: (429, 11)
Test set shape: (185, 11)


<p style="font-size:20px; font-weight:bold;">2.	Model Training (without GridSearchCV)</p>

In [48]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

# Train L1 regularized Logistic Regression (with ALL defaults explicit)
model_l1 = LogisticRegression(
    penalty='l1',
    solver='liblinear',
    C=1.0,                    # Regularization strength (lambda = 1/C)
    fit_intercept=True,       # Automatically learn intercept
    max_iter=100,             # Default iteration limit
    random_state=42
)
model_l1.fit(X_train, y_train)
y_pred_l1 = model_l1.predict(X_test)

print("L1 Accuracy:", accuracy_score(y_test, y_pred_l1))
print(classification_report(y_test, y_pred_l1))

L1 Accuracy: 0.8540540540540541
              precision    recall  f1-score   support

           0       0.94      0.57      0.71        58
           1       0.83      0.98      0.90       127

    accuracy                           0.85       185
   macro avg       0.89      0.78      0.81       185
weighted avg       0.87      0.85      0.84       185



In [34]:
print("=== L1 Model ===")
print(f"Regularization strength: C={model_l1.C}, lambda={1/model_l1.C}")
print(f"Learned coefficients (weights): {model_l1.coef_}")
print(f"Learned intercept: {model_l1.intercept_}")
print(f"Number of features used: {np.sum(model_l1.coef_ != 0)}")

=== L1 Model ===
Regularization strength: C=1.0, lambda=1.0
Learned coefficients (weights): [[-6.95024196e-02  4.15543588e-01  1.66529578e-01 -4.86686929e-01
  -1.11441879e-01  0.00000000e+00 -1.06209050e-01 -3.05998337e-02
   6.17035416e-04  3.19874316e+00  3.32770344e-02]]
Learned intercept: [-2.20404498]
Number of features used: 10


In [47]:
# Train L2 regularized Logistic Regression (with ALL defaults explicit)
model_l2 = LogisticRegression(
    penalty='l2',
    solver='lbfgs',
    C=1.0,                    # Regularization strength (lambda = 1/C)
    fit_intercept=True,       # Automatically learn intercept
    max_iter=1000,            # Explicit iteration limit
    random_state=42
)
model_l2.fit(X_train, y_train)
y_pred_l2 = model_l2.predict(X_test)

print("L2 Accuracy:", accuracy_score(y_test, y_pred_l2))
print(classification_report(y_test, y_pred_l2))

L2 Accuracy: 0.8540540540540541
              precision    recall  f1-score   support

           0       0.94      0.57      0.71        58
           1       0.83      0.98      0.90       127

    accuracy                           0.85       185
   macro avg       0.89      0.78      0.81       185
weighted avg       0.87      0.85      0.84       185



In [35]:
print("\n=== L2 Model ===")
print(f"Regularization strength: C={model_l2.C}, lambda={1/model_l2.C}")
print(f"Learned coefficients (weights): {model_l2.coef_}")
print(f"Learned intercept: {model_l2.intercept_}")
print(f"Number of features used: {np.sum(model_l2.coef_ != 0)}")


=== L2 Model ===
Regularization strength: C=1.0, lambda=1.0
Learned coefficients (weights): [[-0.11008841  0.4843587   0.18447962 -0.51234746 -0.17915165  0.00747102
  -0.11246465 -0.04871727  0.02400291  3.02086854  0.07839731]]
Learned intercept: [-2.10622856]
Number of features used: 11


<p style="font-size:20px; font-weight:bold;">3.	Model Training (with GridSearchCV)</p>

In [43]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression

# Grid search for L1 penalty
param_grid_l1 = {
    'C': [0.01, 0.1, 1, 10, 100],
    'solver': ['liblinear'],
    'penalty': ['l1']
}

logreg_l1 = LogisticRegression(random_state=42, max_iter=1000)

grid_l1 = GridSearchCV(logreg_l1, param_grid_l1, cv=5, scoring='accuracy')
grid_l1.fit(X_train, y_train)

print(" Best params for L1:", grid_l1.best_params_)
print(" Best CV accuracy (L1):", grid_l1.best_score_)

 Best params for L1: {'C': 1, 'penalty': 'l1', 'solver': 'liblinear'}
 Best CV accuracy (L1): 0.7924213406292749


In [44]:
# Predict and evaluate L1
y_pred_l1 = grid_l1.predict(X_test)
print("\n Test accuracy for L1:", accuracy_score(y_test, y_pred_l1))
print(classification_report(y_test, y_pred_l1))


 Test accuracy for L1: 0.8540540540540541
              precision    recall  f1-score   support

           0       0.94      0.57      0.71        58
           1       0.83      0.98      0.90       127

    accuracy                           0.85       185
   macro avg       0.89      0.78      0.81       185
weighted avg       0.87      0.85      0.84       185



In [45]:
# Grid search for L2 penalty
param_grid_l2 = {
    'C': [0.01, 0.1, 1, 10, 100],
    'solver': ['lbfgs'],
    'penalty': ['l2']
}

logreg_l2 = LogisticRegression(random_state=42, max_iter=1000)

grid_l2 = GridSearchCV(logreg_l2, param_grid_l2, cv=5, scoring='accuracy')
grid_l2.fit(X_train, y_train)

print(" Best params for L2:", grid_l2.best_params_)
print(" Best CV accuracy (L2):", grid_l2.best_score_)

 Best params for L2: {'C': 1, 'penalty': 'l2', 'solver': 'lbfgs'}
 Best CV accuracy (L2): 0.7924213406292749


In [46]:
# Predict and evaluate L2
y_pred_l2 = grid_l2.predict(X_test)
print("\n Test accuracy for L2:", accuracy_score(y_test, y_pred_l2))
print(classification_report(y_test, y_pred_l2))


 Test accuracy for L2: 0.8540540540540541
              precision    recall  f1-score   support

           0       0.94      0.57      0.71        58
           1       0.83      0.98      0.90       127

    accuracy                           0.85       185
   macro avg       0.89      0.78      0.81       185
weighted avg       0.87      0.85      0.84       185



**Conclusion:**
Both L1 and L2 regularized Logistic Regression models achieved identical test accuracy (0.854), with high precision and recall for the positive class. Hyperparameter tuning using GridSearchCV did not significantly improve performance, but confirmed the optimal hyperparameters (C = 1) and appropriate solvers (liblinear for L1, lbfgs for L2). This suggests that regularization helped stabilize the model's performance and that default parameters were already near-optimal for this dataset.

**The total performed tests are 4 experiments.**